apps/heft/src/plugins/CopyFilesPlugin.ts

```diff
@@ -0,0 +1,265 @@
+            linkTargetPath: copyDescriptor.sourceFilePath,
+            newLinkPath: copyDescriptor.destinationFilePath
+          });
+          hardlinkCount++;
+        } else {
+          // If it's a copy, simply call the copy function
+          await FileSystem.copyFileAsync({
+            sourcePath: copyDescriptor.sourceFilePath,
+            destinationPath: copyDescriptor.destinationFilePath,
+            alreadyExistsBehavior: AlreadyExistsBehavior.Overwrite
+          });
+          copyCount++;
+        }
+      }
+    );
+
+    return [copyCount, hardlinkCount];
+  }
+
+  private async _getCopyFileDescriptorsAsync(
+    buildFolder: string,
+    copyConfigurations: IExtendedSharedCopyConfiguration[]
+  ): Promise<ICopyFileDescriptor[]> {
+    // Create a map to deduplicate and prevent double-writes
+    // resolvedDestinationFilePath -> [resolvedSourceFilePath, hardlink]
+    const destinationCopyDescriptors: Map<string, ICopyFileDescriptor> = new Map();
+
+    for (const copyConfiguration of copyConfigurations) {
+      // Resolve the source folder path which is where the glob will be run from
+      const resolvedSourceFolderPath: string = path.resolve(buildFolder, copyConfiguration.sourceFolder);
+
+      // Glob extensions with a specific glob to increase perf
+      let sourceFileRelativePaths: Set<string>;
+      if (copyConfiguration.fileExtensions?.length) {
+        const escapedExtensions: string[] = globEscape(copyConfiguration.fileExtensions);
+        const pattern: string = `**/*+(${escapedExtensions.join('|')})`;
+        sourceFileRelativePaths = await this._expandGlobPatternAsync(
+          resolvedSourceFolderPath,
+          pattern,
+          copyConfiguration.excludeGlobs
+        );
+      } else {
+        sourceFileRelativePaths = new Set<string>();
+      }
+
+      // Now include the other glob as well
+      for (const include of copyConfiguration.includeGlobs || []) {
+        const explicitlyIncludedPaths: Set<string> = await this._expandGlobPatternAsync(
+          resolvedSourceFolderPath,
+          include,
+          copyConfiguration.excludeGlobs
+        );
+
+        for (const explicitlyIncludedPath of explicitlyIncludedPaths) {
+          sourceFileRelativePaths.add(explicitlyIncludedPath);
+        }
+      }
+
+      // Dedupe and throw if a double-write is detected
+      for (const destinationFolderRelativePath of copyConfiguration.destinationFolders) {
+        for (const sourceFileRelativePath of sourceFileRelativePaths) {
+          // Only include the relative path from the sourceFolder if flatten is false
+          const resolvedSourceFilePath: string = path.join(resolvedSourceFolderPath, sourceFileRelativePath);
+          const resolvedDestinationFilePath: string = path.resolve(
+            buildFolder,
+            destinationFolderRelativePath,
+            copyConfiguration.flatten ? '.' : path.dirname(sourceFileRelativePath),
```

Comment:
I'm starting to wonder if `flatten` is really a useful option.

Reply:
I still like it in the context of this being a "general purpose" plugin. Even if it's a feature that not many use, it's not exactly much effort to maintain.

Though, if you still hate it, we can pull it out